File: 2018/07/steps.cc
#include <algorithm>
#include <cassert>
#include <chrono>
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>

using namespace std;


class Problem {
public:
	Problem() {};
	Problem(vector<string> lines)
	{
		for (auto line : lines) {
			char	dependency = line[5];
			char	name = line[36];
			this->constraint(name, dependency);
		}
	}

	void
	constraint(char name, char dependency)
	{
		// Make sure both the name and constraint are accounted
		// for. Without this, the starting step won't be recorded.
		if (constraints.count(name) == 0) {
			constraints[name] = set<char>();
		};

		if (constraints.count(dependency) == 0) {
			constraints[dependency] = set<char>();
		}

		constraints[name].insert(dependency);
		// auto temp = constraints[name];
		// temp.insert(dependency);
		// constraints[name] = temp;
	}

	void
	complete(char dependency)
	{
		for (auto step : constraints) {
			if (step.second.count(dependency) == 0) {
				continue;
			}
			constraints[step.first].erase(dependency);
		}
	}

	string
	solve()
	{
		string	result;
		while (constraints.size() > 0) {
			for (auto it : constraints) {
				if (it.second.size() != 0) {
					continue;
				}

				constraints.erase(it.first);
				complete(it.first);
				result.push_back(it.first);
				break;
			}
		}

		return result;
	}

	size_t
	size()
	{
		return this->constraints.size();
	}

private:
	map<char, set<char>>	constraints;
};


vector<string>
readLines(const string path)
{
       ifstream        file(path);
       vector<string>  lines;

       for (string line; getline(file, line); ) {
               lines.push_back(line);
       }
       file.close();

       return lines;
}


const vector<string>	TestSteps = {
	"Step C must be finished before step A can begin.",
	"Step C must be finished before step F can begin.",
	"Step A must be finished before step B can begin.",
	"Step A must be finished before step D can begin.",
	"Step B must be finished before step E can begin.",
	"Step D must be finished before step E can begin.",
	"Step F must be finished before step E can begin."
};


void
self_check_problem_ctor()
{
	Problem		problem(TestSteps);

	assert(problem.size() == 6);
}


void
self_check_solve()
{
	auto expected = "CABDFE";
	Problem		testCase(TestSteps);

	auto result = testCase.solve();
	assert(expected == result);
}


static void
self_check()
{
	self_check_problem_ctor();
	self_check_solve();
	cerr << "self check: OK" << endl;
}


int
main(int argc, char *argv[])
{
	auto start = chrono::system_clock::now();
	self_check();

	for (auto i = 1; i < argc; i++) {
		auto lines = readLines(argv[i]);

		Problem	problem(lines);
		cout << problem.size() << " tasks recorded." << endl;

		auto result = problem.solve();
		cout << "solution: " << result << endl;
	}

	auto finished = chrono::system_clock::now();
	auto span = chrono::duration_cast<chrono::microseconds>(finished - start);
	cout << "solution(s) generated in " << span.count() << "µs" << endl;
}

File: 2018/07/steps.py
#!/usr/bin/env python

import copy
import re
import sys

# { 'a': set('b', 'c') }

STEP_REGEX = re.compile(r"Step (.) must be finished before step (.) can begin.")


def parse_steps(lines):
    steps = {}
    for line in lines:
        m = STEP_REGEX.match(line)
        assert m
        requires, name = m.groups()
        if not name in steps:
            steps[name] = set()
        if not requires in steps:
            steps[requires] = set()
        steps[name].add(requires)

    return steps


TEST_STEPS = {
    "A": set(["C"]),
    "B": set(["A"]),
    "C": set(),
    "D": set(["A"]),
    "E": set(["B", "D", "F"]),
    "F": set(["C"]),
}


def self_check_parse_steps():
    lines = """Step C must be finished before step A can begin.
Step C must be finished before step F can begin.
Step A must be finished before step B can begin.
Step A must be finished before step D can begin.
Step B must be finished before step E can begin.
Step D must be finished before step E can begin.
Step F must be finished before step E can begin.""".split(
        "\n"
    )
    steps = parse_steps(lines)
    assert steps == TEST_STEPS


def mark_done(steps, name):
    for step in steps.keys():
        if name in steps[step]:
            steps[step].remove(name)
    return steps


def sequence(steps):
    result = ""
    while len(steps) > 0:
        for step in sorted(steps.keys()):
            if len(steps[step]) == 0:
                result += step
                del(steps[step])
                steps = mark_done(steps, step)
                break
    return result


def self_check_sequence():
    steps = copy.deepcopy(TEST_STEPS)
    result = sequence(steps)
    assert result == "CABDFE"


def self_check():
    self_check_parse_steps()
    self_check_sequence()

    print('self check OK')

def sequence_file(path):
    with open(path, 'rt') as file:
        return sequence(parse_steps(file))
    
def main(args):
    for path in args:
        print(sequence_file(path))

if __name__ == "__main__":
    self_check()
    main(sys.argv[1:])
